rct229/rules/section5/section5rule21.py

```diff
@@ -0,0 +1,177 @@
+from rct229.ruleset_functions.get_opaque_surface_type import (OpaqueSurfaceType as OST, get_opaque_surface_type)
+from rct229.utils.assertions import getattr_, MissingKeyException
+from rct229.utils.jsonpath_utils import find_all
+from rct229.ruleset_functions.get_surface_conditioning_category_dict import (SurfaceConditioningCategory as SCC, get_surface_conditioning_category_dict)
+from rct229.utils.match_lists import match_lists_exactly_by_id
+
+
+DOOR = schema_enums["SubsurfaceClassificationType"].DOOR.name
+
+
+class Section5Rule21(RuleDefinitionListIndexedBase):
+    """Rule 5 of ASHRAE 90.1-2019 Appendix G Section 5 (Envelope)"""
+
+    def __init__(self):
+        super(Section5Rule21, self).__init__(
+            rmrs_used=UserBaselineProposedVals(False, True, True),
+            required_fields={
+                "$": ["weather"],
+                "weather": ["climate_zone"],
+            },
+            each_rule=Section5Rule21.BuildingRule(),
+            index_rmr="baseline",
+            id="5-21",
+            description="The vertical fenestration shall be distributed on each face of the building in the same proportion as in the proposed design.",
+            list_path="ruleset_model_instances[0].buildings[*]",
+        )
+
+    def create_data(self, context, data=None):
+        rmr_baseline = context.baseline
+        return {"climate_zone": rmr_baseline["weather"]["climate_zone"]}
+
+    class BuildingRule(RuleDefinitionListIndexedBase):
+        def __init__(self):
+            super(Section5Rule21.BuildingRule, self).__init__(
+                rmrs_used=UserBaselineProposedVals(False, True, True),
+                required_fields={},
+                each_rule=Section5Rule21.BuildingRule.AboveGradeWallRule(),
+                index_rmr="baseline",
+            )
+
+        def create_data(self, context, data=None):
+            building_b = context.baseline
+            building_p = context.proposed
+            climate_zone = data["climate_zone"]
+
+            window_wall_areas_dictionary_b = get_area_type_window_wall_area_dict(climate_zone, building_b)
+            window_wall_areas_dictionary_p = get_area_type_window_wall_area_dict(climate_zone, building_p)
+
+            return {
+                **data,
+                "total_fenestration_area_b": sum([window_wall_areas_dictionary_b[key]["total_window_area"]
+                                                  for key in window_wall_areas_dictionary_b.keys()]),
+                "total_fenestration_area_p": sum([window_wall_areas_dictionary_p[key]["total_window_area"]
+                                                  for key in window_wall_areas_dictionary_p.keys()]),
+                "surface_conditioning_category_dict_b": get_surface_conditioning_category_dict(
+                    data["climate_zone"], building_b
+                ),
+            }
+
+        def create_context_list(self, context, data=None):
```

Comment:
Can discuss the implementation. 

I do see a lot this types happening in many rules, but also it is seems like case by case (see rule 5-18)

Reply:
I tried the list_filter method, one issue is we cannot assign the lambda function in the constructor because the data is still not available before calling the create_data function.

It seems to me that if we want to go this route, we need to add an additional function `update_list_filter` to override between lines here: [line](https://github.com/pnnl/ruleset-checking-tool/blob/develop/rct229/rule_engine/rule_base.py#L603-L604)

Reply:
I will move this topic to an issue since this is not a rule-specific request.